src/main/java/seedu/address/logic/commands/GrabCommand.java

```diff
@@ -60,61 +72,101 @@ public CommandResult execute(Model model) throws CommandException {
 
         final StringBuilder builder = new StringBuilder();
 
-        if (index.equals("")) {
-            for (Person p : currAddressBook) {
-                switch (attribute.toString()) {
-                case NAME:
-                    builder.append(p.getName() + "\n");
-                    break;
-                case PHONE:
-                    builder.append(p.getPhone() + "\n");
-                    break;
-                case EMAIL:
-                    builder.append(p.getEmail() + "\n");
-                    break;
-                case ADDRESS:
-                    builder.append(p.getAddress() + "\n");
-                    break;
-                case COURSE:
-                    builder.append(p.getCourse() + "\n");
-                    break;
-                case MATRICCARD:
-                    builder.append(p.getMatricCard() + "\n");
-                    break;
-                case TELEGRAM:
-                    builder.append(p.getTelegram() + "\n");
-                    break;
-                default:
-                }
+        if (tag == null) {
+            // no tag situation
+            if (index.equals("")) {
+                builder.append(grabAllPerson(currAddressBook));
+            } else {
+                int personIndex = Integer.parseInt(index);
+                Person personToBeGrabbed = currAddressBook.get(personIndex - 1);
+                builder.append(grabOnePerson(personToBeGrabbed));
             }
+            return new CommandResult(MESSAGE_SUCCESS + '\n' + builder);
         } else {
-            int personIndex = Integer.parseInt(index);
-            Person personToBeGrabbed = currAddressBook.get(personIndex - 1);
+            if (index.equals("")) {
+                ObservableList<Person> newPersonList =
+                        currAddressBook.filtered(t->t.getTags().contains(tag));
+                if (newPersonList.size() == 0) {
+                    throw new CommandException(MESSAGE_MISSING_TAG);
+                }
+                builder.append(grabAllPerson(newPersonList));
+            } else {
+                // extra index is found
+                throw new CommandException(MESSAGE_GRAB_ONLY_BY_TAG);
+            }
+            return new CommandResult(MESSAGE_SUCCESS + '\n' + builder);
+        }
+    }
+
+    /**
+     * Helper function to grab the person
+     * @param personGetGrabbed the person that get grabbed
+     * @return the grabbed result
+     */
+    public String grabOnePerson(Person personGetGrabbed) {
```

Comment:
i was thinking mebbe `personGrabbed` feels more correct. what you do think?

Reply:
Ok will fix it